src/components/MapboxMap.tsx

```diff
@@ -0,0 +1,163 @@
+  mapboxOptions?: Omit<MapboxOptions, 'container'>,
+  /** Custom Pin component to render for markers on the map. */
+  PinComponent?: PinComponent<T>,
+  /** A function to derive a result's coordinate for the corresponding marker's location on the map. */
+  getCoordinate?: CoordinateGetter<T>,
+  /** A function which is called when user drag the map. */
+  onDrag?: (center: LngLat, bounds: LngLatBounds) => void
+}
+
+/**
+ * A component that renders a map with markers to show result locations using Mapbox GL.
+ *
+ * @remarks
+ * For the map to work properly, be sure to include Mapbox GL stylesheet in the application.
+ *
+ * @example
+ * For instance, user may add the following import statement in their application's index file
+ * or in the file where `MapboxMap` is used:
+ * `import 'mapbox-gl/dist/mapbox-gl.css';`
+ *
+ * Or, user may add a stylesheet link in their html page:
+ * `<link href="https://api.mapbox.com/mapbox-gl-js/VERSION/mapbox-gl.css" rel="stylesheet">`
+ *
+ * @param props - {@link MapboxMapProps}
+ * @returns A React element containing a Mapbox Map
+ *
+ * @public
+ */
+export function MapboxMap<T>({
+  mapboxAccessToken,
+  mapboxOptions,
+  PinComponent,
+  getCoordinate = getDefaultCoordinate,
+  onDrag
+}: MapboxMapProps<T>) {
+  useEffect(() => {
+    mapboxgl.accessToken = mapboxAccessToken;
+  }, [mapboxAccessToken]);
+
+  const mapContainer = useRef<HTMLDivElement>(null);
+  const map = useRef<Map | null>(null);
+  const markers = useRef<Marker[]>([]);
+
+  const locationResults = useSearchState(state => state.vertical.results) as Result<T>[];
+  const onDragDebounced = useDebouncedFunction(onDrag, 100);
+
+  useEffect(() => {
+    if (mapContainer.current && !map.current) {
+      const options: MapboxOptions = {
+        container: mapContainer.current,
+        style: 'mapbox://styles/mapbox/streets-v11',
+        center: [-73.935242, 40.730610],
+        zoom: 9,
+        ...mapboxOptions
+      };
+      map.current = new Map(options);
+      const mapbox = map.current;
+      mapbox.resize();
+      if (onDragDebounced) {
+        mapbox.on('drag', () => {
+          onDragDebounced(mapbox.getCenter(), mapbox.getBounds());
+        });
+      }
+    }
+  }, [mapboxOptions, onDragDebounced]);
+
+  useEffect(() => {
+    markers.current.forEach(marker => marker.remove());
+    markers.current = [];
+    const mapbox = map.current;
+    if (mapbox && locationResults?.length > 0) {
+      const bounds = new LngLatBounds();
+      locationResults.forEach((result, i) => {
+        const markerLocation = getCoordinate(result);
+        if (markerLocation) {
+          const { latitude, longitude } = markerLocation;
+          const el = document.createElement('div');
+          PinComponent && ReactDOM.render(<PinComponent result={result} index={i} />, el);
+          const markerOptions: MarkerOptions = {
+            element: PinComponent ? el : undefined
```

Comment:
instead of checking if the custom PinComponent is specified twice, can the code be wrapped in a single `if` check?

Reply:
I like to have the construction of markerOptions all in one place instead of doing `markerOptions.element = el` separately so I kept it like this. But eh, if check is cool too, I can update

Reply:
I see. I don't have a strong preference, so this is fine